### PR TITLE
convert AudioTrackPlayer

### PR DESCRIPTION
**Pull request by billypchan**
## What's new in this PR?

convert `AudioTrackPlayer` to Swift. Replace key value observation with Notification Center to allow signals transmission form non-objc property to observers.
Refactor the observation of `avPlayer`'s property to prevent overriding `observeValueForKeyPath` which needs manually checking the key path string.

### CODE REVIEW

File: Wire-iOS/Sources/Components/AudioPlayer/AudioTrack.swift [hunk 1]
```diff
@@ -16,15 +16,20 @@
...
-
     var failedToLoad: Bool { get set }
 }
+
+extension AudioTrack {
+    func isEqual(_ other: AudioTrack) -> Bool {
+        return title == title &&
+               author == author &&
+               duration == duration &&
+               streamURL == streamURL
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Components/AudioPlayer/AudioTrack.swift]: I think you've forgotten to use `other`.

Comment (billypchan) [reply Wire-iOS/Sources/Components/AudioPlayer/AudioTrack.swift]: Actually `AudioTrack` is only conform by `ZMAssetClientMessage`. Changed to directly compare the object

File: Wire-iOS/Sources/UserInterface/Conversation/Content/ConversationContentViewController.swift [hunk 2]
```diff
@@ -51,13 +51,13 @@ final class ConversationContentViewController: UIViewController {
...
     var connectionViewController: UserConnectionViewController?
     
-    private var activeMediaPlayerObserver: NSKeyValueObservation?
     private var mediaPlaybackManager: MediaPlaybackManager?
     private var cachedRowHeights: [IndexPath: CGFloat] = [:]
     private var hasDoneInitialLayout = false
     private var onScreen = false
     private weak var messageVisibleOnLoad: ZMConversationMessage?
-    
+    private var token: NSObjectProtocol!
```
Comment (typfel) [inline Wire-iOS/Sources/UserInterface/Conversation/Content/ConversationContentViewController.swift]:
```suggestion
    private var token: NSObjectProtocol?
```

File: Wire-iOS/Sources/Components/AudioPlayer/AudioTrackPlayer.swift [hunk 3]
```diff
@@ -19,14 +19,297 @@
...
+                playerRateObserver = avPlayer?.observe(\AVPlayer.rate, options: [.new]) { [weak self] _, _ in
+                    self?.playRateChanged()
+                }
+
+                playerCurrentItemObserver = avPlayer?.observe(\AVPlayer.currentItem, options: [.new, .initial, .old]) { [weak self] _, _ in
+                    self?.playCurrentItemChanged()
+                }
+            }
+        } else {
+            ///For testing only! streamURL is nil in tests.
```
Comment (typfel) [inline Wire-iOS/Sources/Components/AudioPlayer/AudioTrackPlayer.swift]:
```suggestion
            // For testing only! streamURL is nil in tests.
```

Comment (typfel) [reply Wire-iOS/Sources/Components/AudioPlayer/AudioTrackPlayer.swift]: tripple `///` is only used when documenting classes/properties

File: Wire-iOS/Sources/Managers/MediaPlaybackManager.swift [hunk 4]
```diff
@@ -22,31 +21,32 @@ private let zmLog = ZMSLog(tag: "MediaPlaybackManager")
...
 
 /// This object is an interface for AVS to control conversation media playback
 final class MediaPlaybackManager: NSObject, AVSMedia {
     var audioTrackPlayer: AudioTrackPlayer = AudioTrackPlayer()
-    
-    @objc
-    dynamic private(set) weak var activeMediaPlayer: (MediaPlayer & NSObject)?
-    
+
+    private(set) weak var activeMediaPlayer: MediaPlayer? {
```
Comment (typfel) [inline Wire-iOS/Sources/Managers/MediaPlaybackManager.swift]: So this used to be observed by KVO but is now done through the Notification Center?

File: Wire-iOS/Sources/Components/AudioPlayer/AudioTrackPlayer.swift [hunk 5]
```diff
@@ -19,14 +19,297 @@
...
+
+        if CMTIME_IS_VALID(time) {
+            return TimeInterval(time.value) / TimeInterval(time.timescale)
+        }
+
+        return 0
+    }
+
+    private(set) var playing = false
+    weak var mediaPlayerDelegate: MediaPlayerDelegate?
```
Comment (typfel) [inline Wire-iOS/Sources/Components/AudioPlayer/AudioTrackPlayer.swift]: Let's move this to the other delegate further up?

File: Wire-iOS/Sources/Components/AudioPlayer/AudioTrackPlayer.swift [hunk 6]
```diff
@@ -19,14 +19,297 @@
...
+        guard let time = avPlayer?.currentTime() else { return 0}
+
+        if CMTIME_IS_VALID(time) {
+            return TimeInterval(time.value) / TimeInterval(time.timescale)
+        }
+
+        return 0
+    }
+
+    private(set) var playing = false
```
Comment (typfel) [inline Wire-iOS/Sources/Components/AudioPlayer/AudioTrackPlayer.swift]: Let's move this to the other private properties further up?

File: Wire-iOS/Sources/Managers/MediaPlaybackManager.swift (hunk 4 above)
Comment (billypchan) [reply Wire-iOS/Sources/Managers/MediaPlaybackManager.swift]: Yes, KVO only works on `NSObject`.
The observers( Conversation list, conversation VC) are not tighter link to the `MediaPlayer`

File: Wire-iOS/Sources/Components/AudioPlayer/AudioTrack.swift [hunk 7]
```diff
@@ -16,15 +16,16 @@
...
-    @objc
     var streamURL: URL? { get }
-
     var failedToLoad: Bool { get set }
 }
+
+extension AudioTrack {
+    func isEqual(_ other: AudioTrack) -> Bool {
+        return (self as? NSObject)?.isEqual(other as? NSObject) ?? false
+    }
```
Comment (johnxnguyen) [inline Wire-iOS/Sources/Components/AudioPlayer/AudioTrack.swift]: I'm just curios, why do we need this?

Comment (billypchan) [reply Wire-iOS/Sources/Components/AudioPlayer/AudioTrack.swift]: `AudioTrack` was a objc protocol which has build-in `isEqual` method. In`isOwnTrackPlayingInAudioPlayer`, we have to check the playing track is the same as message's track.

Comment (typfel) [reply Wire-iOS/Sources/Components/AudioPlayer/AudioTrack.swift]: So we could inherit from `NSObjectProtocol` and delete this method? You are already implicitly relying on it here by casting to `NSObject`.

Comment (billypchan) [reply Wire-iOS/Sources/Components/AudioPlayer/AudioTrack.swift]: fixed

**Pull request merged**